daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -36,7 +45,21 @@ private[speedy] sealed abstract class SBuiltin(val arity: Int) {
   private[speedy] def execute(args: util.ArrayList[SValue], machine: Machine): Unit
 }
 
+private[speedy] sealed abstract class SBuiltinPure(val arity1: Int) extends SBuiltin(arity1) {
+
+  override private[speedy] final def execute(
+      args: util.ArrayList[SValue],
+      machine: Machine): Unit = {
+    machine.returnValue = executePure(args)
+  }
+
+  /** Execute the (pure) builtin with 'arity' number of arguments in 'args'.
+    Returns the resultin value */
+  private[speedy] def executePure(args: util.ArrayList[SValue]): SValue
```

Comment:
```suggestion
  protected def executePure(args: util.ArrayList[SValue]): SValue
```

Reply:
We can't make this change, because we need `executePure` to be called from `SExpr` (in the coming ANF PR).